README.rst

```diff
@@ -0,0 +1,806 @@
+Context
+=======
+
+We build and release software by massively consuming and producing software
+packages such as NPMs, RPMs, Rubygems, etc.
+
+Each package manager, platform, type or ecosystem has its own conventions and
+protocols to identify, locate and provision software packages.
+
+
+Problem
+=======
+
+When tools, APIs and databases process or store multiple package types, it is
+difficult to reference the same software package across tools in a uniform way.
+
+For example, these tools, specifications and API use relatively similar
+approaches to identify and locate software packages, each with subtle
+differences in syntax, naming and conventions:
+
+- Grafeas uses a scheme, namespace, name and version in a URL-like string.
+  https://github.com/Grafeas/Grafeas
+
+- Here.com OSRK uses a package manager, name and version field and a colon-
+  separated URL-like string
+  https://github.com/heremaps/oss-review-toolkit
+
+- JFrog XRay uses a scheme, namespace, name and version in a URL-like string
+  https://www.jfrog.com/confluence/display/XRAY/Xray+REST+API#XrayRESTAPI-ComponentIdentifiers
+
+- Libraries.io uses a platform, name and version
+  https://libraries.io/
+
+- OpenShift fabric8 analytics uses ecosystem, name and version
+  https://github.com/fabric8-analytics/
+
+- ScanCode and AboutCode.org use a type, name and version
+  https://github.com/nexB/scancode-toolkit
+
+- SPDX has an appendix for external repository references and uses a type and a
+  locator with a type-specific syntax for component separators in a URL-like
+  string
+  https://github.com/spdx/spdx-spec/blob/master/chapters/3-package-information.md
+
+- versioneye uses a type, name and version
+  https://github.com/versioneye/
+
+Solution
+========
+
+A `purl` or package URL is an attempt to standardize existing approaches to
+reliably identify and locate software packages.
+
+A `purl` is a URL string used to identify and locate a software package in a
+mostly universal and uniform way across programing languages, package managers,
+packaging conventions, tools, APIs and databases.
+
+Such a package URL is useful to reliably reference the same software package
+using a simple and expressive syntax and conventions based on familiar URLs.
+
+
+purl
+~~~~~
+
+`purl` stands for **package URL**.
+
+A `purl` is a URL composed of six components::
+
+    type:namespace/name@version?qualifiers#subpath
+
+Components are separated by a specific character for unambiguous parsing.
+
+The defintion for each components is:
+
+- **type**: the package "type" or package "protocol" such as maven, npm, nuget,
+  gem, pypi, etc. Required.
+- **namespace**: some name prefix such as a Maven groupid, a Docker image owner,
+  a GitHub user or organization. Optional and type-specific.
```

Comment:
Just gonna add it here too: the npm version of this starts with `@`. It would be super nice to not have to escape that. 😆

Reply:
@iarna I am with you there, but FWIW there are some percent-encoding alright in NPMs registry API URLs:
For instance https://registry.yarnpkg.com/@invisionag%2feslint-config-ivx/ works but https://registry.yarnpkg.com/@invisionag/feslint-config-ivx/  comes out as `{"error":"Not found"}` just so you known.... the slash between the "namespace" and name must be percent encoded....and yes, I know you might think that I have researched too many of these tiny details and quirks. 

Here in a `purl` I guess this is not strictly needed: a name is always required. Therefore a leading `@` in a scoped NPM `namespace` is never ambiguous. What would be ambiguous is if we allow unescaped `@` anywhere in the name or namespace and a `purl` comes with no version.

For instance, say a package name is `super@package` for the hypothetical `weirdo` package type:

With the `purl` `weirdo:super@package` I would parse `@``package` as a version even though no version was provided: `weirdo:super@package@1.2.3` would not be ambiguous though of course, but my hypothetical weirdo packages rarely have a version attached...

So for simplicity I specified that the whole name and each namespace segment should be percent-encoded: we could relax this for the leading `@` in namespace/name (to make scoped NPMs look beautiful) alright. I will update this and both can be acceptable in any case.